File: src/test/java/cube8540/oauth/authentication/credentials/oauth/security/endpoint/OAuth2TokenIntrospectionEndpointTest.java
package cube8540.oauth.authentication.credentials.oauth.security.endpoint;

import cube8540.oauth.authentication.credentials.oauth.error.InvalidRequestException;
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2AccessTokenDetails;
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2AccessTokenDetailsService;
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2ClientDetails;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.security.authentication.InsufficientAuthenticationException;
import org.springframework.security.oauth2.core.OAuth2Error;
import org.springframework.security.oauth2.core.OAuth2ErrorCodes;

import java.security.Principal;
import java.util.Map;

import static cube8540.oauth.authentication.credentials.oauth.security.endpoint.TokenEndpointTestHelper.RAW_TOKEN_ID;
import static cube8540.oauth.authentication.credentials.oauth.security.endpoint.TokenEndpointTestHelper.makeAccessTokenDetails;
import static cube8540.oauth.authentication.credentials.oauth.security.endpoint.TokenEndpointTestHelper.makeAccessTokenDetailsService;
import static cube8540.oauth.authentication.credentials.oauth.security.endpoint.TokenEndpointTestHelper.makeAccessTokenMap;
import static cube8540.oauth.authentication.credentials.oauth.security.endpoint.TokenEndpointTestHelper.makeClientCredentialsToken;
import static cube8540.oauth.authentication.credentials.oauth.security.endpoint.TokenEndpointTestHelper.makeClientDetails;
import static cube8540.oauth.authentication.credentials.oauth.security.endpoint.TokenEndpointTestHelper.makeEmptyAccessTokenDetailsService;
import static cube8540.oauth.authentication.credentials.oauth.security.endpoint.TokenEndpointTestHelper.makeIntrospectionConverter;
import static cube8540.oauth.authentication.credentials.oauth.security.endpoint.TokenEndpointTestHelper.makeNotClientCredentialsTokenPrincipal;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

@DisplayName("토큰 정보 확인 엔드 포인트 테스트")
class OAuth2TokenIntrospectionEndpointTest {

    @Test
    @DisplayName("요청 정보에 토큰이 null 일때 토큰 정보 검색")
    void searchTokenInfoWhenRequestTokenNull() {
        Principal principal = makeClientCredentialsToken(null);
        OAuth2AccessTokenDetailsService service = makeEmptyAccessTokenDetailsService();
        OAuth2TokenIntrospectionEndpoint endpoint = new OAuth2TokenIntrospectionEndpoint(service);

        OAuth2Error error = assertThrows(InvalidRequestException.class, () -> endpoint.introspection(principal, null)).getError();
        assertEquals(OAuth2ErrorCodes.INVALID_REQUEST, error.getErrorCode());
    }

    @Test
    @DisplayName("인증 객체의 타입이 ClientCredentialsToken이 아닐때 토큰 정보 검색")
    void searchTokenInfoWhenPrincipalObjectTypeNotClientCredentialsToken() {
        Principal principal = makeNotClientCredentialsTokenPrincipal();
        OAuth2AccessTokenDetailsService service = makeEmptyAccessTokenDetailsService();
        OAuth2TokenIntrospectionEndpoint endpoint = new OAuth2TokenIntrospectionEndpoint(service);

        assertThrows(InsufficientAuthenticationException.class, () -> endpoint.introspection(principal, RAW_TOKEN_ID));
    }

    @Test
    @DisplayName("인증의 상세 정보가 OAuth2ClientDetails가 이닐때 토큰 정보 검색")
    void searchTokenInfoWhenTokenDetailsTypeNotOauth2ClientDetails() {
        Principal principal = makeClientCredentialsToken(new Object());
        OAuth2AccessTokenDetailsService service = makeEmptyAccessTokenDetailsService();
        OAuth2TokenIntrospectionEndpoint endpoint = new OAuth2TokenIntrospectionEndpoint(service);

        assertThrows(InsufficientAuthenticationException.class, () -> endpoint.introspection(principal, RAW_TOKEN_ID));
    }

    @Test
    @DisplayName("액세스 토큰 정보 검색")
    void searchTokenInfo() {
        OAuth2ClientDetails clientDetails = makeClientDetails();
        Principal principal = makeClientCredentialsToken(clientDetails);
        OAuth2AccessTokenDetails accessToken = makeAccessTokenDetails();
        OAuth2AccessTokenDetailsService service = makeAccessTokenDetailsService(RAW_TOKEN_ID, accessToken);
        Map<String, Object> accessTokenMap = makeAccessTokenMap();
        OAuth2AccessTokenIntrospectionConverter converter = makeIntrospectionConverter(accessToken, accessTokenMap);
        OAuth2TokenIntrospectionEndpoint endpoint = new OAuth2TokenIntrospectionEndpoint(service);

        endpoint.setConverter(converter);

        assertEquals(accessTokenMap, endpoint.introspection(principal, RAW_TOKEN_ID));
    }

    @Test
    @DisplayName("인증 객체의 타입이 ClientCredentialsToken이 아닐때 토큰 소유자 검색")
    void searchTokenOwnerWhenPrincipalTypeNotClientCredentialsToken() {
        Principal principal = makeNotClientCredentialsTokenPrincipal();
        OAuth2AccessTokenDetailsService service = makeEmptyAccessTokenDetailsService();
        OAuth2TokenIntrospectionEndpoint endpoint = new OAuth2TokenIntrospectionEndpoint(service);

        assertThrows(InsufficientAuthenticationException.class, () -> endpoint.userInfo(principal, RAW_TOKEN_ID));
    }

    @Test
    @DisplayName("요청 받은 토큰이 null 일때 토큰 소유자 검색")
    void searchTokenOwnerWhenRequestTokenIsNull() {
        Principal principal = makeClientCredentialsToken(null);
        OAuth2AccessTokenDetailsService service = makeEmptyAccessTokenDetailsService();
        OAuth2TokenIntrospectionEndpoint endpoint = new OAuth2TokenIntrospectionEndpoint(service);

        OAuth2Error error = assertThrows(InvalidRequestException.class, () -> endpoint.userInfo(principal, null)).getError();
        assertEquals(OAuth2ErrorCodes.INVALID_REQUEST, error.getErrorCode());
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/oauth/security/endpoint/OAuth2TokenIntrospectionEndpointTest.kt
package cube8540.oauth.authentication.oauth.security.endpoint

import cube8540.oauth.authentication.oauth.error.InvalidRequestException
import cube8540.oauth.authentication.oauth.security.OAuth2AccessTokenDetails
import cube8540.oauth.authentication.oauth.security.OAuth2AccessTokenDetailsService
import cube8540.oauth.authentication.oauth.security.OAuth2ClientDetails
import cube8540.oauth.authentication.oauth.security.provider.ClientCredentialsToken
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.catchThrowable
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.springframework.security.authentication.InsufficientAuthenticationException
import org.springframework.security.oauth2.core.OAuth2ErrorCodes
import java.security.Principal

class OAuth2TokenIntrospectionEndpointTest {

    private val converter: OAuth2AccessTokenIntrospectionConverter = mockk()
    private val tokenDetailsService: OAuth2AccessTokenDetailsService = mockk()

    private val endpoint = OAuth2TokenIntrospectionEndpoint(tokenDetailsService)

    init {
        endpoint.converter = converter
    }

    @Nested
    inner class ReadTokenTest {

        @Test
        fun `requested principal class type is not credentials client`() {
            val principal: Principal = mockk()

            val thrown = catchThrowable { endpoint.introspection(principal, "token") }
            assertThat(thrown).isInstanceOf(InsufficientAuthenticationException::class.java)
        }

        @Test
        fun `requested principal details class type is not oauth2 client details`() {
            val principal: ClientCredentialsToken = mockk {
                every { principal } returns mockk<Principal>()
            }

            val thrown = catchThrowable { endpoint.introspection(principal, "token") }
            assertThat(thrown).isInstanceOf(InsufficientAuthenticationException::class.java)
        }

        @Test
        fun `requested token value is null`() {
            val principal: ClientCredentialsToken = mockk {
                every { principal } returns mockk<OAuth2ClientDetails>()
            }

            val thrown = catchThrowable { endpoint.introspection(principal, null) }
            assertThat(thrown).isInstanceOf(InvalidRequestException::class.java)
            assertThat((thrown as InvalidRequestException).error.errorCode).isEqualTo(OAuth2ErrorCodes.INVALID_REQUEST)
        }

        @Test
        fun `read token`() {
            val clientDetails: OAuth2ClientDetails = mockk()
            val principal: ClientCredentialsToken = mockk {
                every { principal } returns clientDetails
            }
            val accessToken: OAuth2AccessTokenDetails = mockk()
            val convertedAccessTokenMap: Map<String, String?> = mockk()

            every { tokenDetailsService.readAccessToken("tokenId") } returns accessToken
            every { converter.convertAccessToken(accessToken) } returns convertedAccessTokenMap

            val result = endpoint.introspection(principal, "tokenId")
            assertThat(result).isEqualTo(convertedAccessTokenMap)
        }
    }
}